### PR TITLE
Feat/social auth

### PR DESCRIPTION
**Pull request by moo-gh**

### CODE REVIEW

File: df_auth/serializers.py [hunk 1]
```diff
@@ -92,3 +94,30 @@ def validate(self, attrs):
...
         return attrs
+
+
+class OAuth2InputSerializer(serializers.Serializer):
+    provider = serializers.CharField(required=False)
+    code = serializers.CharField()
+    redirect_uri = serializers.CharField(required=False)
+
+
+class TokenSerializer(serializers.Serializer):
```
Comment (ayushin) [inline df_auth/serializers.py]: This should not be needed as we should re-use TokenObtainSerializer, perhaps by splitting out AbstractTokenSerialiser first

File: df_auth/viewsets.py [hunk 2]
```diff
@@ -52,3 +73,38 @@ def blacklist(self, request, *args, **kwargs):
 class OTPViewSet(ValidationOnlyCreateViewSet):
     serializer_class = OTPObtainSerializer
     permission_classes = (permissions.AllowAny,)
+
+
+class SignIn(GenericAPIView):
```
Comment (ayushin) [inline df_auth/viewsets.py]: We should follow the pattern of using ViewSets and router

File: df_auth/viewsets.py [hunk 3]
```diff
@@ -52,3 +73,38 @@ def blacklist(self, request, *args, **kwargs):
...
+    serializer_class = JWTPairSerializer
+
+    def get_object(self):
```
Comment (ayushin) [inline df_auth/viewsets.py]: Without going into understanding what it does, can the authentication logic be inside of serialiser validate and authentication backend to follow the same pattern we use in this module?

File: df_auth/viewsets.py [hunk 4]
```diff
@@ -52,3 +73,38 @@ def blacklist(self, request, *args, **kwargs):
...
+        return user
+
+    def post(self, request, provider):
+        decorate_request(request, provider)
+        OAuth2InputSerializer(data=self.request.data).is_valid(raise_exception=True)
+        try:
+            user = self.get_object()
+        except (AuthException, HTTPError) as e:
+            logger.error(e)
+            return response.Response("something wrong happened", status.HTTP_400_BAD_REQUEST)
```
Comment (ayushin) [inline df_auth/viewsets.py]: this will break openapi schema

File: df_auth/viewsets.py [hunk 5]
```diff
@@ -52,3 +73,38 @@ def blacklist(self, request, *args, **kwargs):
...
+class Connect(SignIn):
+    def post(self, request, provider):
+        if not request.user.is_authenticated:
+            raise NotAuthenticated()
+        return super().post(request, provider)
+
+
+class CallBack(APIView):
+    def get(self, request):
+        return response.Response({'code': request.GET['code']})
```
Comment (ayushin) [inline df_auth/viewsets.py]: this will also break openapi schema

**Review by ayushin — CHANGES_REQUESTED** [review]

File: df_auth/viewsets.py (hunk 4 above)
Comment (moo-gh) [reply df_auth/viewsets.py]: How should I resolve this?

File: df_auth/viewsets.py (hunk 5 above)
Comment (moo-gh) [reply df_auth/viewsets.py]: I guess it is resolved by using a serializer for output. I can see this API in the swagger.

File: df_auth/viewsets.py (hunk 2 above)
Comment (moo-gh) [reply df_auth/viewsets.py]: Done, used router for social.

File: df_auth/serializers.py (hunk 1 above)
Comment (moo-gh) [reply df_auth/serializers.py]: Done, used TokenObtainSerializer for generating access token by doing some minor change on it.

File: df_auth/viewsets.py [hunk 6]
```diff
@@ -52,3 +70,40 @@ def blacklist(self, request, *args, **kwargs):
...
+    def get_object(self):
+        user = self.request.user
+        self.request.backend.redirect_uri = settings.REST_SOCIAL_OAUTH_REDIRECT
+        user = user_is_authenticated(user) and user or None
+        self.request.backend.STATE_PARAMETER = False
+        user = self.request.backend.complete(user=user)
+        return user
+
+    @action(methods=["post"], detail=False)
+    def signin(self, request):
```
Comment (ayushin) [inline df_auth/viewsets.py]: why is it no possible to have this logic in serializer validate following the same pattern of the simplejwt?

Is it really necessary to override get_object? This does not feel intuitive.

Comment (moo-gh) [reply df_auth/viewsets.py]: Not resolved yet. I'm still working on that 🙂

File: df_auth/viewsets.py (hunk 3 above)
Comment (moo-gh) [reply df_auth/viewsets.py]: Done, moved to the serializer validate function.

File: df_auth/viewsets.py (hunk 6 above)
Comment (moo-gh) [reply df_auth/viewsets.py]: Done, removed get_object function and used create functions as we create tokens.

File: df_auth/settings.py [hunk 7]
```diff
@@ -3,7 +3,7 @@
 
 
 DEFAULTS = {
-    "USER_IDENTITY_FIELDS": ("email",),
+    "USER_IDENTITY_FIELDS": ("email","username"),
```
Comment (ayushin) [inline df_auth/settings.py]: is this still needed?

File: df_auth/social_auth.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+from social_core.utils import get_strategy
+from social_django.strategy import DjangoStrategy
+from social_django.utils import psa
+from social_django.utils import STORAGE
+
+
+class DRFStrategy(DjangoStrategy):
+    def __init__(self, storage, request=None, tpl=None):
+        self.request = request
+        super(DjangoStrategy, self).__init__(storage, tpl)
```
Comment (ayushin) [inline df_auth/social_auth.py]: this is not pythonic 3 super should be without class

File: df_auth/viewsets.py [hunk 9]
```diff
@@ -52,3 +55,28 @@ def blacklist(self, request, *args, **kwargs):
 class OTPViewSet(ValidationOnlyCreateViewSet):
     serializer_class = OTPObtainSerializer
     permission_classes = (permissions.AllowAny,)
+
+
+class SocialAuth(viewsets.GenericViewSet):
+    serializer_class = SocialTokenObtainSerializer
+
+    def create(self, request, *args, **kwargs):
```
Comment (ayushin) [inline df_auth/viewsets.py]: this should be DRY, i.e. inherit from the Mixin and you don't need create method at all

**Comment by ayushin** [comment]
I added more comments

File: df_auth/serializers.py [hunk 10]
```diff
@@ -92,3 +100,31 @@ def validate(self, attrs):
...
+    provider = serializers.CharField(write_only=True)
+    code = serializers.CharField(write_only=True)
+
+
+class SocialTokenObtainSerializer(
+    AbstractTokenSerializer, AbstractSocialIdentitySerializer
+):
+    def validate(self, attrs):
+        request = self.context["request"]
+        user = request.user
```
Comment (ayushin) [inline df_auth/serializers.py]: this should not be needed at all after you split theserializer into TokenObtain and Connect

File: df_auth/settings.py (hunk 7 above)
Comment (moo-gh) [reply df_auth/settings.py]: No, it's not required for social auth. I added it to get the access token with a username and password as I didn't succeed to do that with email and password. I guess authenticate function works with username and password. 🤔

File: df_auth/social_auth.py (hunk 8 above)
Comment (moo-gh) [reply df_auth/social_auth.py]: Done.👌

File: df_auth/viewsets.py (hunk 9 above)
Comment (moo-gh) [reply df_auth/viewsets.py]: Done, used the ValidationOnlyCreateViewSet.

File: df_auth/serializers.py (hunk 10 above)
Comment (moo-gh) [reply df_auth/serializers.py]: I created a specific serializer for SocialConnect and did some changes to them. 👍

**Comment by ayushin** [comment]
This is how we wanted this done:

https://github.com/djangoflow/django-df-auth/pull/6

**Pull request closed without merging**